### PR TITLE
Feature/command group cache 2 btc 576

### PR DESCRIPTION
**Pull request by blackyblack**

### CODE REVIEW

File: include/veriblock/command_group_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  virtual void clear() {
+    _cache.clear();
+    _keys.clear();
+    _priority.clear();
+  }
+
+ protected:
+  size_t _maxsize;
+  std::list<id_t> _priority;
```
Comment (Warchant) [inline include/veriblock/command_group_cache.hpp:64]: use std::vector

File: include/veriblock/command_group_cache.hpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  std::unordered_map<id_t, std::list<id_t>::iterator> _keys;
```
Comment (Warchant) [inline include/veriblock/command_group_cache.hpp:65]:
```suggestion
  std::unordered_map<id_t, std::vector<id_t>::iterator> _keys;
```

Comment (Warchant) [reply include/veriblock/command_group_cache.hpp:65]: std::list is very slow, instead would be better to use something like circular buffer built on top of std::vector

File: include/veriblock/blockchain/vbk_block_addon.hpp [hunk 3]
```diff
@@ -27,6 +27,12 @@ struct VbkBlockAddon :
 
   bool payloadsIdsEmpty() const { return vtbids.empty(); }
 
+  template <typename pop_t, typename pop_id_t>
+  std::vector<pop_id_t>& getPayloadIds();
```
Comment (Mr-Leshiy) [inline include/veriblock/blockchain/vbk_block_addon.hpp:31]: Why is it so important to specify this method with the pop_t type ?  Why does not enough only pop_id_t ?

File: include/veriblock/storage/payloads_storage.hpp [hunk 4]
```diff
@@ -65,44 +77,120 @@ class PayloadsStorage {
...
+    return _cgValidity.find(cid) != _cgValidity.end();
+  }
+
  protected:
   std::shared_ptr<PayloadsRepository<ATV>> _repoAtv;
   std::shared_ptr<PayloadsRepository<VTB>> _repoVtb;
   std::shared_ptr<PayloadsRepository<VbkBlock>> _repoBlocks;
+  CommandGroupCache _cacheAlt;
+  CommandGroupCache _cacheVbk;
+  std::unordered_map<CommandGroup::id_t, bool> _cgValidity;
```
Comment (Warchant) [inline include/veriblock/storage/payloads_storage.hpp:178]: Shouldn't it be `std::vector, bool`?
where vector is `containing block hash + command group id` ?

Comment (blackyblack) [reply include/veriblock/storage/payloads_storage.hpp:178]: Yes. But it doesn't matter since CommandGroup::id_t is a vector itself.

**Pull request merged**